plugins/grid-bookmark/src/GridBookmarkWidget/components/Highlight/Highlight.tsx

```diff
@@ -0,0 +1,129 @@
+import React from 'react'
+import { observer } from 'mobx-react'
+import { makeStyles } from 'tss-react/mui'
+import { getSession, isSessionModelWithWidgets } from '@jbrowse/core/util'
+import { LinearGenomeViewModel } from '@jbrowse/plugin-linear-genome-view'
+
+// icons
+import BookmarkIcon from '@mui/icons-material/Bookmark'
+
+// locals
+import { Tooltip } from '@mui/material'
+import { GridBookmarkModel } from '../../model'
+
+type LGV = LinearGenomeViewModel
+
+const useStyles = makeStyles()({
+  highlight: {
+    height: '100%',
+    position: 'absolute',
+    textAlign: 'center',
+    overflow: 'hidden',
+    display: 'flex',
+    alignItems: 'start',
+  },
+})
+
+interface BookmarkInfo {
+  width: number
+  left: number
+  highlight: string
+  label: string
+}
+
+interface LocalStorageBookmark {
+  start: number
+  end: number
+  assemblyName: string
+  refName: string
+  reversed: boolean
+  label: string
+  highlight: string
+}
+
+const intensify = (rgba: string, opacity: number) => {
+  if (rgba) {
+    const values = rgba.replaceAll(/[^\d,.]/g, '').split(',')
+    const originalOpacity = values.pop()
+    // originalOpacity !== '0' assumes if the user has set the opacity of this highlight to 0, they probably don't want to see the label either
+    const n = `rgba(${values.join(', ')}, ${
+      originalOpacity !== '0' ? opacity : 0
+    })`
+    return n
+  }
+  return `rgba(0,0,0,0)`
+}
+
+const Highlight = observer(function Highlight({ model }: { model: LGV }) {
+  const { classes } = useStyles()
+
+  const session = getSession(model)
+  if (!isSessionModelWithWidgets(session)) {
+    return null
+  }
+
+  let bookmarkWidget = session.widgets.get('GridBookmark') as GridBookmarkModel
+  if (!bookmarkWidget) {
```

Comment:
I think it might be worth avoiding doing 'side effects' like addWidget in the body of the react render (e.g. move to useEffect if needed? if its needed at all?)

Reply:
It's needed to access the bookmarks without retrieving directly from the localstorage (i.e. it's going to be using the existing grabs within the bookmarks model instead of re-writing them all)

the motivation here was basically, if you're viewing highlights you're probably going to want to be using the bookmarks widget anyways

i moved it into a useState ? not sure if this resolves your concern but lemme know if you have a better soln for this guy

Reply:
reverted cus the linter reminded me why i did it like this in the first place:

`React Hook "useState" is called conditionally. React Hooks must be called in the exact same order in every component render` thoughts?

Reply:
move to a useEffect instead of a useState. useEffect is a good way of expressing that "you are doing something as a sideneffect of rendering" which in this case, as part of rendering this component, you are opening the bookmark widget.

Reply:
also, the linter/runtime warning there is correct: gotta follow the rules of hooks https://legacy.reactjs.org/docs/hooks-rules.html (or new docs https://react.dev/warnings/invalid-hook-call-warning)

Reply:
@cmdcolin lemme know how that looks in the most recent commit